### PR TITLE
How-To: Monitor PVC Usage

### PR DESCRIPTION
**Pull request by bastjan**
## Checklist

- [x] Try to isolate changes into separate PRs (to build a better changelog).
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `how-to`, `tutorial`, `reference`, `explanation`, `cicd`, `dependency`
      as they show up in the changelog

### CODE REVIEW

File: docs/modules/ROOT/pages/how-to/monitor-pvc-usage.adoc [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+kubelet_volume_stats_used_bytes
+
+kubelet_volume_stats_inodes
+kubelet_volume_stats_inodes_free
+kubelet_volume_stats_inodes_used
+----
+
+== Prerequisites
+
+* Command Line Access to APPUiO Cloud
```
Comment (simu) [inline docs/modules/ROOT/pages/how-to/monitor-pvc-usage.adoc:22]:
```suggestion
* Command Line Access to {product}
```

File: docs/modules/ROOT/pages/how-to/monitor-pvc-usage.adoc [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+== Create a Disk Usage Alert
+
+The following instructions create two alerts that trigger when the disk usage or used https://en.wikipedia.org/wiki/Inode[inodes] of a PVC exceed 90%.
```
Comment (simu) [inline docs/modules/ROOT/pages/how-to/monitor-pvc-usage.adoc:26]:
```suggestion
The following instructions create two alerts that trigger when the disk usage or used https://en.wikipedia.org/wiki/Inode[inodes] of any PVC in a namespace exceed 90%.
```

File: docs/modules/ROOT/pages/how-to/monitor-pvc-usage.adoc [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+spec:
+  groups:
+  - name: my-app-volume.rules
+    rules:
+    - alert: VolumeAvailableSpaceLow
+      expr: |
+        (
+          sum(kubelet_volume_stats_available_bytes{persistentvolumeclaim!=""}) by (persistentvolumeclaim)
+        /
+          sum(kubelet_volume_stats_capacity_bytes{persistentvolumeclaim!=""}) by (persistentvolumeclaim)
```
Comment (simu) [inline docs/modules/ROOT/pages/how-to/monitor-pvc-usage.adoc:53]: the `sum by (persistentvolumeclaim) ()` are technically not necessary, but probably don't hurt.

File: docs/modules/ROOT/pages/how-to/monitor-pvc-usage.adoc [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+          Your application may become unable to write data to disk.
+
+          Either increase the PVC size or delete files to free up space.
+    - alert: VolumeFreeInodesLow
+      expr: |
+        (
+          sum(kubelet_volume_stats_inodes_free{persistentvolumeclaim!=""}) by (persistentvolumeclaim)
+        /
+          sum(kubelet_volume_stats_inodes{persistentvolumeclaim!=""}) by (persistentvolumeclaim)
```
Comment (simu) [inline docs/modules/ROOT/pages/how-to/monitor-pvc-usage.adoc:71]: same here

**Review by simu — CHANGES_REQUESTED** [review]
I'm missing documentation on routing alerts to e-mail (or Slack, or somewhere).

I'm fine with that part being a separate how-to, but that how-to should be linked from all the "setting up prometheusrules" how-tos.

**Comment by bastjan** [comment]
> I'm missing documentation on routing alerts to e-mail (or Slack, or somewhere).

Will add in a separate PR...

File: docs/modules/ROOT/pages/how-to/monitor-pvc-usage.adoc (hunk 3 above)
Comment (bastjan) [reply docs/modules/ROOT/pages/how-to/monitor-pvc-usage.adoc:53]: I'm pretty sure it's needed if volume changes between nodes and there's some overlap due to the scrape interval.

Maybe it should be `max/min` though.

Also i need to test RWX volumes...

**Review by akosma — APPROVED** [review]

File: docs/modules/ROOT/pages/how-to/monitor-pvc-usage.adoc (hunk 3 above)
Comment (bastjan) [reply docs/modules/ROOT/pages/how-to/monitor-pvc-usage.adoc:53]: > I'm pretty sure it's needed if volume changes between nodes and there's some overlap due to the scrape interval.

The query is not failing but can trigger N alerts for each mounted node.

**Comment by bastjan** [comment]
Ceph backed volumes don't have "real" inodes and report zero inodes free. Not sure yet how to handle this since we can't differentiate those pvcs.

**Comment by simu** [comment]
> Ceph backed volumes don't have "real" inodes and report zero inodes free. Not sure yet how to handle this since we can't differentiate those pvcs.

You can get the storage class for each pvc with something like

```
kubelet_volume_stats_inodes *on(persistentvolumeclaim) group_left(storageclass) kube_persistentvolumeclaim_info
```

File: docs/modules/ROOT/pages/how-to/monitor-pvc-usage.adoc (hunk 3 above)
Comment (simu) [reply docs/modules/ROOT/pages/how-to/monitor-pvc-usage.adoc:53]: > > I'm pretty sure it's needed if volume changes between nodes and there's some overlap due to the scrape interval.
> 
> The query is not failing but can trigger N alerts for each mounted node.

right, but then using `sum()` would also report wrong absolute numbers for RWX volumes which are mounted on multiple nodes, so I think using `max()` would make more sense

**Comment by bastjan** [comment]
> I assume the inode issue is only with CephFS volumes.

This is correct.

I've added a filter by `storageclass` as  suggested since we control the storage class names on APPUiO Cloud.

**Review by simu — APPROVED** [review]
LGTM

**Review by corvus-ch — CHANGES_REQUESTED** [review]
There is a discussion suggesting to use `max without` which is not yet taken into account (but was marked resolved?).

Also, using static usage percentage is likely to produce noise. Instead, I would prefer to showcase `predict_linear`. Alert, if the disk will run full in a given time period based on current growth.

Also, we need to wait with the merge until this is actually available on the system. Alternatively, we need to place a disclaimer.

**Comment by bastjan** [comment]
> Also, using static usage percentage is likely to produce noise. Instead, I would prefer to showcase predict_linear. Alert, if the disk will run full in a given time period based on current growth.

Not sure it that is a "save" or comprehensible default. Would adding it as a tip be ok?

File: docs/modules/ROOT/pages/how-to/monitor-pvc-usage.adoc (hunk 3 above)
Comment (bastjan) [reply docs/modules/ROOT/pages/how-to/monitor-pvc-usage.adoc:53]: The supplied config is correct as `sum() / sum()` is fine. The tip is also correct but can trigger multiple alerts.

I'll add `max without(instance,node)` to the absolute value tip.

**Comment by bastjan** [comment]
@corvus-ch I've added a section about prediction alerts.

**Review by corvus-ch — APPROVED** [review]

**Pull request merged**